include/native_thread_pool.h

```diff
@@ -0,0 +1,220 @@
+                return seastar::make_exception_future<>(
+                  std::move(state).get_exception());
+            } else {
+                return seastar::make_ready_future<>();
+            }
+        });
+    }
+};
+
+struct SubmitQueue {
+    seastar::semaphore free_slots;
+    seastar::gate pending_tasks;
+    explicit SubmitQueue(size_t num_free_slots)
+      : free_slots(num_free_slots) {
+    }
+    seastar::future<> stop() {
+        return pending_tasks.close();
+    }
+};
+
+/// an engine for scheduling non-seastar tasks from seastar fibers
+class ThreadPool {
+    std::atomic<bool> stopping = false;
+    std::mutex mutex;
+    std::condition_variable cond;
+    std::vector<std::thread> threads;
+    seastar::sharded<SubmitQueue> submit_queue;
+    const size_t queue_size;
+    boost::lockfree::queue<WorkItem*> pending;
+    semaphore add_task_sem;
+
+    void loop() {
+        for (;;) {
+            WorkItem* work_item = nullptr;
+            {
+                std::unique_lock lock{mutex};
+                cond.wait_for(
+                  lock, std::chrono::milliseconds(10), [this, &work_item] {
+                      return pending.pop(work_item) || is_stopping();
+                  });
+            }
+            if (work_item) {
+                work_item->process();
+            } else if (is_stopping()) {
+                break;
+            }
+        }
+    }
+    bool is_stopping() const {
+        return stopping.load(std::memory_order_relaxed);
+    }
+    static void pin(unsigned cpu_id) {
+        cpu_set_t cs;
+        CPU_ZERO(&cs);
+        CPU_SET(cpu_id, &cs);
+        [[maybe_unused]] auto r = pthread_setaffinity_np(
+          pthread_self(), sizeof(cs), &cs);
+    }
+    seastar::semaphore& local_free_slots() {
+        return submit_queue.local().free_slots;
+    }
+    ThreadPool(const ThreadPool&) = delete;
+    ThreadPool& operator=(const ThreadPool&) = delete;
+
+public:
+    /**
+     * @param queue_sz the depth of pending queue. before a task is scheduled,
+     *                 it waits in this queue. we will round this number to
+     *                 multiple of the number of cores.
+     * @param n_threads the number of threads in this thread pool.
+     * @param cpu the CPU core to which this thread pool is assigned
+     * @note each @c Task has its own crimson::thread::Condition, which
+     * possesses possesses an fd, so we should keep the size of queue under a
+     * reasonable limit.
+     */
+    ThreadPool(size_t n_threads, size_t queue_size, unsigned cpu_id)
+      : queue_size{queue_size} // round_up_to(queue_sz, seastar::smp::count)}
+      , pending{queue_size}
+      , add_task_sem(n_threads) {
+        for (size_t i = 0; i < n_threads; i++) {
+            threads.emplace_back([this, cpu_id] {
+                pin(cpu_id);
+                loop();
+            });
+        }
+    }
+    ~ThreadPool() {
+        for (auto& thread : threads) {
+            thread.join();
```

Comment:
Is it destroyed on seastar thread?

Reply:
Yes

Reply:
Got it, so it's not good then. It might introduce reactor stall on shutdown. Maybe we should create detached threads so no implicit join would be needed? We can busy-wait in 'stop' method to ensure that the threads are stopped. Or even do nothing.
No need to change this now I think.